### DISCUSSION TITLE
Open a text file for reading and writing

### DISCUSSION THREAD

**Discussion by MaxGyver83** [janet-lang/janet#1413] [Q&A] [answered]
When I open a text file in mode `:w`, `:a` or `:w+`, it appears to be empty:

```janet
(def filehandle (file/open "/tmp/file" :w))
  (var line nil)
  (while true
    (set line (file/read filehandle :line))
    (print (if line (string "line: " line) "Empty file!"))
    (unless line (break)))
(file/close filehandle)
```

Even worse: After trying to print all lines, the file content is deleted. (Only with `:w` and `:w+`. `:a` does not delete on opening/reading.)

Isn't it possible to open a file for reading **and** writing?

Do I have to open a file twice in such a case? If I want to read and write multiple times in a loop, can I keep the file open twice (for reading and for writing)?

**Comment by ianthehenry** [comment, marked as answer]
`file/open` is a wrapper around `fopen`, so `:w` and `:w+` truncate the file as usual. If you want to open for reading and writing without truncating the file, use `:r+`.

Reading and writing the same file is a little bit cursed so it might be worth talking about the thing you're trying to do. It's usually better to read the whole file, edit its contents in memory, then write the new result out to a brand new file, and then `os/rename` it to overwrite the original file.

You can also use use `:a+`, but note that the file is positioned at the end initially, so you have to seek to the beginning (`(file/seek filehandle :set 0)`) before you can read anything out of it. And that doesn't let you reposition your writes using seek; all writes will always append to the end of the file. Which might be appropriate for what you're trying to do.

**Reply by MaxGyver83** [reply]
Thanks a lot! `:r+` is exactly what I need. This terminology is really confusing: `:a` stands for *append*. `+` also stands for *append*. So I couldn't have guessed what `:a+` does. :smile:

I think I would write in the API explicitly that `:w` means **over**write.

For now, I changed my code to first *slurp* the file and then opening it for appending only. (I'll share a link when it's online.)

**Reply by sogaiu** [reply]
FYI: https://github.com/janet-lang/janet/issues/1205
